ingestion-beam/src/main/java/com/mozilla/telemetry/io/Read.java

```diff
@@ -56,4 +67,69 @@ public FileInput(ValueProvider<String> fileSpec, InputFileFormat fileFormat) {
       return input.apply(TextIO.read().from(fileSpec)).apply(fileFormat.decode());
     }
   }
+
+  /** Implementation of reading from local or remote files. */
+  public static class BigQueryInput extends Read {
+
+    private final ValueProvider<String> tableSpec;
+    private final BigQueryReadMethod method;
+    private final Source source;
+    private final String rowRestriction;
+    private final List<String> selectedFields;
+
+    public enum Source {
+      TABLE, QUERY
+    }
+
+    public BigQueryInput(ValueProvider<String> tableSpec, BigQueryReadMethod method, Source source,
+        String rowRestriction, List<String> selectedFields) {
+      this.tableSpec = tableSpec;
+      this.method = method;
+      this.source = source;
+      this.rowRestriction = rowRestriction;
+      this.selectedFields = selectedFields;
+    }
+
+    @Override
+    public Result<PCollection<PubsubMessage>> expand(PBegin input) {
+      BigQueryIO.TypedRead<PubsubMessage> read = BigQueryIO
+          .read((SchemaAndRecord schemaAndRecord) -> {
+            TableSchema tableSchema = schemaAndRecord.getTableSchema();
+            GenericRecord record = schemaAndRecord.getRecord();
+            byte[] payload = ((ByteBuffer) record.get("payload")).array();
+
+            // We copy only string-type fields to attributes, which is complete for raw and
+            // error tables; decoded payload tables have the nested metadata object also encoded in
+            // the payload, so we can safely drop the metadata object here and rely on ParsePayload
+            // to parse attributes from it.
```

Comment:
> rely on ParsePayload to parse attributes from it.

but the decoder runs `ParseUri` before `ParsePayload` and throws `NullUriException` if the `uri` attribute is missing. should we attempt to extract just `uri` from the `metadata` object?

Reply:
Good catch!

We still have the issue of BQ streaming inserts failing and getting put to error output, in which case we lose all attributes, and we only have the metadata to refer to. So I think we need to solve that case as well. It's not immediately clear to me how best to handle that.

Reply:
Perhaps we should remove the `NullUriException` and allow message to pass through `ParseUri` unchanged in case the "uri" attribute is missing. Then, in ParsePayload, after the payload is parsed and we call `AddMetadata.stripPayloadMetadataToAttributes`, we should be good.  We could add a sanity check at that point to make sure that doctype and namespace are present and throw an exception if missing.

Reply:
> We still have the issue of BQ streaming inserts failing and getting put to error output, in which case we lose all attributes, and we only have the metadata to refer to. So I think we need to solve that case as well. It's not immediately clear to me how best to handle that.

for streaming inserts we get errors, could we restore `uri` when we [turn the row back into a pubsub message](https://github.com/mozilla/gcp-ingestion/blob/master/ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java#L410-L433)?

for file loads do we just...not get errors? I don't see error handling after [insert](https://github.com/mozilla/gcp-ingestion/blob/master/ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java#L446-L450), so I assume they crash the job?

Reply:
> for file loads do we just...not get errors? I don't see error handling after insert, so I assume they crash the job?

Yes, it's my expectation that they would crash the job and bubble up.

Reply:
> Perhaps we should remove the `NullUriException` ... Then, in `ParsePayload`, after the payload is parsed and we call `AddMetadata.stripPayloadMetadataToAttributes`, we should be good.

the reason we throw `NullUriException` is we don't want to gunzip & json parse if we don't know what schema to validate.

Reply:
but actually, i think if `uri` is missing, this is fine, because we know we did that